File: backend/src/main/java/com/myfinance/backend/service/InvestmentService.java
package com.myfinance.backend.service;

import com.myfinance.backend.model.Investment;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface InvestmentService {

    List<Investment> getAllInvestments();

    Optional<Investment> getInvestmentById(UUID id);

    Investment createInvestment(Investment investment);

    Investment updateInvestment(UUID id, Investment investment);

    void deleteInvestment(UUID id);

    List<Investment> getInvestmentsByType(String type);

    List<Investment> getInvestmentsByTicker(String ticker);

    List<Investment> getInvestmentsByCurrentValueGreaterThan(Double amount);

    List<Investment> getInvestmentsLastSyncedBefore(LocalDateTime dateTime);

    /** Syncs crypto prices for the current household's ticker+quantity investments. */
    int syncPrices();

    /** Same as {@link #syncPrices()} but across every household - the daily scheduler's job. */
    int syncPricesForAllHouseholds();

    /**
     * Records a new purchase: blends into the existing quantity/amountInvested (both are
     * simply summed, which is exactly the weighted-average-cost formula) rather than
     * requiring the caller to hand-compute new totals.
     */
    Investment addPurchase(UUID id, BigDecimal quantity, BigDecimal unitPrice);
}

File: frontend/src/service/InvestmentService.ts
import type { AxiosError } from 'axios';
import api from '../config/axios';
import type { Investment } from '../model/InvestmentModel';

const ENDPOINT = '/investments';

export const getAllInvestments = async (): Promise<Investment[]> => {
  const response = await api.get<Investment[]>(ENDPOINT);
  return response.data;
};

export const getInvestmentById = async (id: string): Promise<Investment | null> => {
  try {
    const response = await api.get<Investment>(`${ENDPOINT}/${id}`);
    return response.data;
  } catch (error: unknown) {
    const axiosError = error as AxiosError;
    if (axiosError?.response?.status === 404) return null;
    throw error;
  }
};

export const createInvestment = async (investment: Investment): Promise<Investment> => {
  const response = await api.post<Investment>(ENDPOINT, investment);
  return response.data;
};

export const updateInvestment = async (id: string, investment: Investment): Promise<Investment> => {
  const response = await api.put<Investment>(`${ENDPOINT}/${id}`, investment);
  return response.data;
};

export const deleteInvestment = async (id: string): Promise<void> => {
  await api.delete(`${ENDPOINT}/${id}`);
};

export const getInvestmentsByType = async (type: string): Promise<Investment[]> => {
  const response = await api.get<Investment[]>(`${ENDPOINT}/type/${type}`);
  return response.data;
};

export const getInvestmentsByTicker = async (ticker: string): Promise<Investment[]> => {
  const response = await api.get<Investment[]>(`${ENDPOINT}/ticker/${ticker}`);
  return response.data;
};

export const getInvestmentsByCurrentValueGreaterThan = async (amount: number): Promise<Investment[]> => {
  const response = await api.get<Investment[]>(`${ENDPOINT}/value-greater-than`, {
    params: { amount },
  });
  return response.data;
};

export const getInvestmentsLastSyncedBefore = async (dateTime: string): Promise<Investment[]> => {
  const response = await api.get<Investment[]>(`${ENDPOINT}/last-synced-before`, {
    params: { dateTime },
  });
  return response.data;
};
